File: Sprite.java
/*
The MIT License (MIT)

Copyright (c) 2005 Alex Duesel, http://www.mandarine.,tv

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

import javax.microedition.lcdui.*;

 class Sprite  {
	 
     static PushBoxCanvas engine;
	
	 static final int 
	 ////////////////////
	 SPEED = 4,
	 TILE_STEPS_DY = 12,
	 TILE_STEPS = TILE_STEPS_DY,
	 TILE_X = PushBoxCanvas.TILE_DX>>1,
	 TILE_Y = PushBoxCanvas.TILE_DY>>1,
	 STATE_IDLE			= 0,
	 STATE_MOVING		= 1,
	 STATE_PUSHED		= 2,
	 NUM_DIRECTIONS	= 4,
	 DIR_NONE	= -1,
	 DIR_NORTH	= 0,
	 DIR_EAST	= 1,
	 DIR_SOUTH	= 2,
	 DIR_WEST	= 3,
	 DIR_X[] = {0,1,0,-1},
	 DIR_Y[] = {-1,0,1,0},
	 DIR_TILE_X[] = {2,2,-2,-2},
	 DIR_TILE_Y[] = {-1,1,1,-1},
	 BOX			= 0,
	 TARGET			= 1,
	 BOX2		= 2,
	 DEFAULT_FRAME_RATE = 150;	
	
	 static final byte
	 ////////////////////
	 SPRITE_WIDTH[] = { 27,45,45},
	 SPRITE_HEIGHT[] = { 45,44,44},
	 SPRITE_SCAN[] = {1,1,1},
	 SPRITE_OFFSET_X[] = {-14,-21,-21},
	 SPRITE_OFFSET_Y[] = {-41,-32,-32},
	 FRAMES_MOVING[] = {1,0,2,0};
		
	 int
	 //////////////////////
	 frame,
	 frameIndex,
	 frameRate,
	 frameDelay,
	 loops,
	 globalIndex,
	 localIndex,		
	 type,
	 state,
	 ticks,
	 x,
	 y,
	 height,
	 dir,
	 speed,
	 currentTile,
	 tileRow,
	 tileX,
	 tileY,
	 tileOffset,
	 tileTicks,
	 frameOffset,
	 walk,
	 steps,
	 swidth,
	 sheight,
	 scanHeight,
	 scanMask,
	 offsetX,
	 offsetY;
	
	 byte 
	 //////////////
	 frames[];
	
	 boolean
	 ///////////////////
	 isPlayer,
	 moved,
	 painted,
	 move;

	 Image
	 //////////////
	 image;

	 Sprite 
	 /////////////////
	 pushing,
	 prev,
	 next;
		 	
	 Sprite(int globalIndex) {
		this.globalIndex = globalIndex;
	 }
	
	 
	 void init(int type) {
		this.type = type;
		isPlayer = (type==BOX);
		walk = DIR_NONE;
		steps = 0;
		pushing = null;
		move = false;
		setState(STATE_IDLE);		
		Image[] imgs = new Image[]{PushBoxCanvas.data[5],PushBoxCanvas.data[2],PushBoxCanvas.data[4]};		
		setSpriteImage(imgs[type],type);		
	}
	
	 void setState(int state) {		
		frameOffset = 0;	
		byte[] frames = null;
		int frameRate = DEFAULT_FRAME_RATE;
		switch(state) {
			case STATE_IDLE:
				if (isPlayer) {
					frameOffset+=(dir*3);
				}
				break;
			case STATE_MOVING:
				if (isPlayer) {
					frameOffset+=(dir*3);
					frames = FRAMES_MOVING;
				}
				break;
			}
		reset(frames,frameRate);
		this.state = state;
	}
	
	 void setTile(int tileX, int tileY,int dir) {
		tileRow = tileX+tileY;
		x = engine.tile_x(tileX,tileY)+TILE_X;
		y = engine.tile_y(tileX,tileY)+TILE_Y;		
		currentTile = engine.getTile(tileX,tileY);		
		height = 0;		
		this.dir = dir;		
		this.tileX = tileX;
		this.tileY = tileY;		
		steps = 0;		
		tileOffset = 0;
		tileTicks = 0;		
		moved = true;		
		setState(STATE_IDLE);
	}
	
	 boolean doesOccupy(int tileX, int tileY) {
		boolean doesOccupy = false;
		int tx = this.tileX;
		int ty = this.tileY;
		if (tx==tileX && ty==tileY) {
			doesOccupy = true;
		} else if (tileOffset!=0) {
			if (this.tileX+DIR_X[dir]==tx && this.tileY+DIR_Y[dir]==ty) {
				tx-=DIR_X[dir];	
				ty-=DIR_Y[dir];	
			} else {
				tx+=DIR_X[dir];	
				ty+=DIR_Y[dir];	
			}
			if (tx==tileX && ty==tileY)
				doesOccupy = true;
		}
		return doesOccupy;
	}
	
	 void update(int frameTime) {
		boolean stop;
		int distance;
		if (ticks>0) {			
			ticks--;			
			if (ticks==0) {
				setState(state);
			}
		}
		switch(state) {
			case STATE_IDLE:
				if (isPlayer) {
					if (move) {
						move = false;
						if (move(dir)) {
							update(frameTime);
							return;
						}
					}
					if (walk!=DIR_NONE) {
						if (move(walk)) {
							update(frameTime);
							return;
						} else {
							dir = walk;
							setState(STATE_IDLE);
						}
					}						
				}	
				if (frames==FRAMES_MOVING)
					setState(STATE_IDLE);
				break;
			case STATE_MOVING:
				stop = false;
				distance = SPEED;
				if (steps<distance) {
					stop = true;
					if (isPlayer) {
						boolean next = (walk==dir);
						if (next) {
							if (move(dir))
								stop = false;
						}
					} 
					if (stop) distance = steps;
				}
				steps-=distance;
				move(dir,distance);
				if (stop) {
					stop();
					state = STATE_IDLE;
				}
				break;
			}
		tileTicks++;
		animate(frameTime);
		walk = DIR_NONE;
	}
	
	 boolean move(int dir) {
		boolean move = true;
		boolean set = true;
		move = canMove(dir,0);
		if (this.dir==dir && state==STATE_MOVING)
			set = false;	
		if (move || isPlayer)
			this.dir = dir;
		if (move) {
			this.move = false;
			push(dir);
			steps+= TILE_STEPS_DY;
			if (set)
				setState(STATE_MOVING);
			else
				state = STATE_MOVING;
		}
		if(engine.demo)engine.demo_idx++;

		return move;
	}
	
	 void push(int dir) {
		this.dir = dir;
		if (pushing==null)	
			pushing = engine.getObjectSuperType(this.tileX+DIR_X[dir],this.tileY+DIR_Y[dir]);
		if (pushing!=null) {
			pushing.push(dir);
			if (engine.getTile(pushing.tileX+DIR_X[dir],pushing.tileY+DIR_Y[dir])==1) {
				pushing.image=engine.data[4];
				if (engine.getTile(pushing.tileX,pushing.tileY)!=1)
				engine.target_cnt+=1;
				engine.notifyt(this, STATE_PUSHED);
			} else if (engine.getTile(pushing.tileX+DIR_X[dir],pushing.tileY+DIR_Y[dir])==0) {
				if(pushing.image==engine.data[4])
					engine.target_cnt-=1;
				pushing.image=engine.data[2];
			}
			pushing.setState(STATE_PUSHED);
		}
	}
	
	 void stop() {
		if (pushing!=null) {
			pushing.stop();
			pushing = null;
		}
		if (state==STATE_PUSHED)
			state = STATE_IDLE;
		
		steps = 0;
		tileOffset = 0;
	}
	
	 boolean canMove(int dir,int stack) {
		if (!engine.tile_accessible(this,tileX+DIR_X[dir],tileY+DIR_Y[dir],dir))
			return false;
		boolean canMove = true;
		Sprite block =  engine.getObjectSuperType(tileX+DIR_X[dir],tileY+DIR_Y[dir]);	
		if (block!=null) {
			canMove = false;
			if (block!=pushing && (steps!=0||block.state!=STATE_IDLE)) {
			} else if (stack<1) {
				if (block.type==TARGET ||block.type== BOX2) {
					if (isPlayer)
						canMove = block.canMove(dir,stack++);
				}	
			}
		}
		return canMove;
	}
	
	 void changeTile() {
		tileX+=DIR_X[dir];
		tileY+=DIR_Y[dir];
		currentTile = engine.getTile(tileX,tileY);
		tileTicks = 0;
	}
	
	 void move(int dir, int distance) {			
		int offset = tileOffset+distance;		
		if (offset>=TILE_STEPS) {
			tileRow = tileX+tileY;
			tileOffset-=TILE_STEPS;
		}
		if ((tileOffset==0 || offset>TILE_STEPS)&&(dir==DIR_SOUTH || dir==DIR_EAST))
			tileRow = tileX+DIR_X[dir]+tileY+DIR_Y[dir];
		tileOffset+=distance;		
		if ((tileX!=tileX+DIR_X[dir] || tileY!=tileY+DIR_Y[dir]) && tileOffset>=(TILE_STEPS>>1))
			changeTile();		
		x+=DIR_TILE_X[dir]*distance;
		y+=DIR_TILE_Y[dir]*distance;		
		if (pushing!=null)
			pushing.move(dir,distance);
		moved = true;
	}
		
	 void setKeys(int dir) {		
		walk = dir;
	}
		
	 void setSpriteImage(Image img, int type) {
		 image = img;
		 swidth=SPRITE_WIDTH[type];
		 sheight=SPRITE_HEIGHT[type];
		 scanHeight=SPRITE_SCAN[type];
			for (int s=scanHeight ; --s>=0 ; ) 
				scanMask |= (0x1<<s);
		offsetX = SPRITE_OFFSET_X[type];
		offsetY = SPRITE_OFFSET_Y[type];		
	}
	
	 void reset(byte[] frames,int frameRate) {
		this.frames = frames;
		this.frameRate = frameRate;	
		frameIndex = 0;
		frameDelay = frameRate;
		if (frames==null)
			frame = 0;
		else
			frame = frames[frameIndex];	
		loops = 0;
		painted = false;
	}
	
	 void animate(int frameTime) {
		if (frameTime>frameRate) frameTime = frameRate;
		frameDelay-=frameTime;
		if (!painted) return;
		if (frames!=null && frameDelay<=0) {				
			frameIndex++;				
			if (frameIndex==frames.length) {
				frameIndex = 0;
				loops++;
			}				
			frameDelay+=frameRate;
			frame = frames[frameIndex];
			painted = false;
		}
	}
	 
	 void paint(Graphics g) {
			int xframe = frameOffset+this.frame;
			paint(g,x,y-height,xframe);
			painted = true;
		}
		
	void paint(Graphics g, int x, int y, int xframe) {
			x += offsetX;
			y += offsetY;
			int frameX = (xframe >> scanHeight)*swidth;
			int frameY = (xframe & scanMask)*sheight;
            g.setClip(x,y,swidth,sheight);
			g.drawImage(image,x-frameX,y-frameY,Graphics.TOP | Graphics.LEFT);
		}
}

File: web/js/Sprite.js
/*
The MIT License (MIT)

Copyright (c) 2005 Alex Duesel, http://www.mandarine.tv
HTML5 / JavaScript conversion (c) 2025

  Sprite.js  --  direct port of the 2005 Sprite.java.
  The class structure, constants, field list and the movement / pushing
  logic are kept 1:1 with the original Java so the old code stays
  recognizable; only the syntax is now JavaScript.
*/

"use strict";

class Sprite {

  constructor(globalIndex) {
    // --- instance fields (same names as the Java source) ---
    this.frame = 0; this.frameIndex = 0; this.frameRate = 0; this.frameDelay = 0;
    this.loops = 0; this.globalIndex = globalIndex; this.localIndex = 0;
    this.type = 0; this.state = 0; this.ticks = 0;
    this.x = 0; this.y = 0; this.height = 0; this.dir = 0; this.speed = 0;
    this.currentTile = 0; this.tileRow = 0; this.tileX = 0; this.tileY = 0;
    this.tileOffset = 0; this.tileTicks = 0; this.frameOffset = 0;
    this.walk = 0; this.steps = 0;
    this.swidth = 0; this.sheight = 0; this.scanHeight = 0; this.scanMask = 0;
    this.offsetX = 0; this.offsetY = 0;
    this.frames = null;
    this.isPlayer = false; this.moved = false; this.painted = false; this.move = false;
    this.image = null;
    this.pushing = null; this.prev = null; this.next = null;
  }

  init(type) {
    this.type = type;
    this.isPlayer = (type === Sprite.BOX);
    this.walk = Sprite.DIR_NONE;
    this.steps = 0;
    this.pushing = null;
    this.move = false;
    this.setState(Sprite.STATE_IDLE);
    // original referenced the static PushBoxCanvas.data[]; here data lives on
    // the engine instance, so we read it through Sprite.engine.
    const imgs = [Sprite.engine.data[5], Sprite.engine.data[2], Sprite.engine.data[4]];
    this.setSpriteImage(imgs[type], type);
  }

  setState(state) {
    this.frameOffset = 0;
    let frames = null;
    let frameRate = Sprite.DEFAULT_FRAME_RATE;
    switch (state) {
      case Sprite.STATE_IDLE:
        if (this.isPlayer) {
          this.frameOffset += (this.dir * 3);
        }
        break;
      case Sprite.STATE_MOVING:
        if (this.isPlayer) {
          this.frameOffset += (this.dir * 3);
          frames = Sprite.FRAMES_MOVING;
        }
        break;
    }
    this.reset(frames, frameRate);
    this.state = state;
  }

  setTile(tileX, tileY, dir) {
    this.tileRow = tileX + tileY;
    this.x = Sprite.engine.tile_x(tileX, tileY) + Sprite.TILE_X;
    this.y = Sprite.engine.tile_y(tileX, tileY) + Sprite.TILE_Y;
    this.currentTile = Sprite.engine.getTile(tileX, tileY);
    this.height = 0;
    this.dir = dir;
    this.tileX = tileX;
    this.tileY = tileY;
    this.steps = 0;
    this.tileOffset = 0;
    this.tileTicks = 0;
    this.moved = true;
    this.setState(Sprite.STATE_IDLE);
  }

  doesOccupy(tileX, tileY) {
    let doesOccupy = false;
    let tx = this.tileX;
    let ty = this.tileY;
    if (tx === tileX && ty === tileY) {
      doesOccupy = true;
    } else if (this.tileOffset !== 0) {
      if (this.tileX + Sprite.DIR_X[this.dir] === tx && this.tileY + Sprite.DIR_Y[this.dir] === ty) {
        tx -= Sprite.DIR_X[this.dir];
        ty -= Sprite.DIR_Y[this.dir];
      } else {
        tx += Sprite.DIR_X[this.dir];
        ty += Sprite.DIR_Y[this.dir];
      }
      if (tx === tileX && ty === tileY)
        doesOccupy = true;
    }
    return doesOccupy;
  }

  update(frameTime) {
    let stop;
    let distance;
    if (this.ticks > 0) {
      this.ticks--;
      if (this.ticks === 0) {
        this.setState(this.state);
      }
    }
    switch (this.state) {
      case Sprite.STATE_IDLE:
        if (this.isPlayer) {
          if (this.move) {
            this.move = false;
            if (this.moveDir(this.dir)) {
              this.update(frameTime);
              return;
            }
          }
          if (this.walk !== Sprite.DIR_NONE) {
            if (this.moveDir(this.walk)) {
              this.update(frameTime);
              return;
            } else {
              this.dir = this.walk;
              this.setState(Sprite.STATE_IDLE);
            }
          }
        }
        if (this.frames === Sprite.FRAMES_MOVING)
          this.setState(Sprite.STATE_IDLE);
        break;
      case Sprite.STATE_MOVING:
        stop = false;
        distance = Sprite.SPEED;
        if (this.steps < distance) {
          stop = true;
          if (this.isPlayer) {
            const next = (this.walk === this.dir);
            if (next) {
              if (this.moveDir(this.dir))
                stop = false;
            }
          }
          if (stop) distance = this.steps;
        }
        this.steps -= distance;
        this.moveStep(this.dir, distance);
        if (stop) {
          this.stop();
          this.state = Sprite.STATE_IDLE;
        }
        break;
    }
    this.tileTicks++;
    this.animate(frameTime);
    this.walk = Sprite.DIR_NONE;
  }

  // Java had two overloads of move(); JS has no overloading, so the
  // single-arg "try to move in dir" version is moveDir() and the
  // (dir,distance) translation version is moveStep().
  moveDir(dir) {
    let move = true;
    let set = true;
    move = this.canMove(dir, 0);
    if (this.dir === dir && this.state === Sprite.STATE_MOVING)
      set = false;
    if (move || this.isPlayer)
      this.dir = dir;
    if (move) {
      this.move = false;
      this.push(dir);
      this.steps += Sprite.TILE_STEPS_DY;
      if (set)
        this.setState(Sprite.STATE_MOVING);
      else
        this.state = Sprite.STATE_MOVING;
    }
    if (Sprite.engine.demo) Sprite.engine.demo_idx++;
    return move;
  }

  push(dir) {
    this.dir = dir;
    if (this.pushing == null)
      this.pushing = Sprite.engine.getObjectSuperType(this.tileX + Sprite.DIR_X[dir], this.tileY + Sprite.DIR_Y[dir]);
    if (this.pushing != null) {
      this.pushing.push(dir);
      if (Sprite.engine.getTile(this.pushing.tileX + Sprite.DIR_X[dir], this.pushing.tileY + Sprite.DIR_Y[dir]) === 1) {
        this.pushing.image = Sprite.engine.data[4];
        if (Sprite.engine.getTile(this.pushing.tileX, this.pushing.tileY) !== 1)
          Sprite.engine.target_cnt += 1;
        Sprite.engine.notifyt(this, Sprite.STATE_PUSHED);
      } else if (Sprite.engine.getTile(this.pushing.tileX + Sprite.DIR_X[dir], this.pushing.tileY + Sprite.DIR_Y[dir]) === 0) {
        if (this.pushing.image === Sprite.engine.data[4])
          Sprite.engine.target_cnt -= 1;
        this.pushing.image = Sprite.engine.data[2];
      }
      this.pushing.setState(Sprite.STATE_PUSHED);
    }
  }

  stop() {
    if (this.pushing != null) {
      this.pushing.stop();
      this.pushing = null;
    }
    if (this.state === Sprite.STATE_PUSHED)
      this.state = Sprite.STATE_IDLE;
    this.steps = 0;
    this.tileOffset = 0;
  }

  canMove(dir, stack) {
    if (!Sprite.engine.tile_accessible(this, this.tileX + Sprite.DIR_X[dir], this.tileY + Sprite.DIR_Y[dir], dir))
      return false;
    let canMove = true;
    const block = Sprite.engine.getObjectSuperType(this.tileX + Sprite.DIR_X[dir], this.tileY + Sprite.DIR_Y[dir]);
    if (block != null) {
      canMove = false;
      if (block !== this.pushing && (this.steps !== 0 || block.state !== Sprite.STATE_IDLE)) {
        // blocked
      } else if (stack < 1) {
        if (block.type === Sprite.TARGET || block.type === Sprite.BOX2) {
          if (this.isPlayer)
            canMove = block.canMove(dir, stack++);
        }
      }
    }
    return canMove;
  }

  changeTile() {
    this.tileX += Sprite.DIR_X[this.dir];
    this.tileY += Sprite.DIR_Y[this.dir];
    this.currentTile = Sprite.engine.getTile(this.tileX, this.tileY);
    this.tileTicks = 0;
  }

  moveStep(dir, distance) {
    const offset = this.tileOffset + distance;
    if (offset >= Sprite.TILE_STEPS) {
      this.tileRow = this.tileX + this.tileY;
      this.tileOffset -= Sprite.TILE_STEPS;
    }
    if ((this.tileOffset === 0 || offset > Sprite.TILE_STEPS) && (dir === Sprite.DIR_SOUTH || dir === Sprite.DIR_EAST))
      this.tileRow = this.tileX + Sprite.DIR_X[dir] + this.tileY + Sprite.DIR_Y[dir];
    this.tileOffset += distance;
    if ((this.tileX !== this.tileX + Sprite.DIR_X[dir] || this.tileY !== this.tileY + Sprite.DIR_Y[dir]) && this.tileOffset >= (Sprite.TILE_STEPS >> 1))
      this.changeTile();
    this.x += Sprite.DIR_TILE_X[dir] * distance;
    this.y += Sprite.DIR_TILE_Y[dir] * distance;
    if (this.pushing != null)
      this.pushing.moveStep(dir, distance);
    this.moved = true;
  }

  setKeys(dir) {
    this.walk = dir;
  }

  setSpriteImage(img, type) {
    this.image = img;
    this.swidth = Sprite.SPRITE_WIDTH[type];
    this.sheight = Sprite.SPRITE_HEIGHT[type];
    this.scanHeight = Sprite.SPRITE_SCAN[type];
    for (let s = this.scanHeight; --s >= 0;)
      this.scanMask |= (0x1 << s);
    this.offsetX = Sprite.SPRITE_OFFSET_X[type];
    this.offsetY = Sprite.SPRITE_OFFSET_Y[type];
  }

  reset(frames, frameRate) {
    this.frames = frames;
    this.frameRate = frameRate;
    this.frameIndex = 0;
    this.frameDelay = frameRate;
    if (frames == null)
      this.frame = 0;
    else
      this.frame = frames[this.frameIndex];
    this.loops = 0;
    this.painted = false;
  }

  animate(frameTime) {
    if (frameTime > this.frameRate) frameTime = this.frameRate;
    this.frameDelay -= frameTime;
    if (!this.painted) return;
    if (this.frames != null && this.frameDelay <= 0) {
      this.frameIndex++;
      if (this.frameIndex === this.frames.length) {
        this.frameIndex = 0;
        this.loops++;
      }
      this.frameDelay += this.frameRate;
      this.frame = this.frames[this.frameIndex];
      this.painted = false;
    }
  }

  paint(g) {
    const xframe = this.frameOffset + this.frame;
    this.paintFrame(g, this.x, this.y - this.height, xframe);
    this.painted = true;
  }

  paintFrame(g, x, y, xframe) {
    x += this.offsetX;
    y += this.offsetY;
    const frameX = (xframe >> this.scanHeight) * this.swidth;
    const frameY = (xframe & this.scanMask) * this.sheight;
    g.setClip(x, y, this.swidth, this.sheight);
    g.drawImage(this.image, x - frameX, y - frameY, Graphics.TOP | Graphics.LEFT);
  }
}

/* ---- static constants (verbatim from Sprite.java) ------------------------ */
Sprite.engine = null;

Sprite.SPEED = 4;
Sprite.TILE_STEPS_DY = 12;
Sprite.TILE_STEPS = Sprite.TILE_STEPS_DY;
// TILE_X / TILE_Y reference PushBoxCanvas tile size (set after that class loads)
Sprite.STATE_IDLE = 0;
Sprite.STATE_MOVING = 1;
Sprite.STATE_PUSHED = 2;
Sprite.NUM_DIRECTIONS = 4;
Sprite.DIR_NONE = -1;
Sprite.DIR_NORTH = 0;
Sprite.DIR_EAST = 1;
Sprite.DIR_SOUTH = 2;
Sprite.DIR_WEST = 3;
Sprite.DIR_X = [0, 1, 0, -1];
Sprite.DIR_Y = [-1, 0, 1, 0];
Sprite.DIR_TILE_X = [2, 2, -2, -2];
Sprite.DIR_TILE_Y = [-1, 1, 1, -1];
Sprite.TYPE_LION = 0;      // the player (== BOX index in the image table)
Sprite.BOX = 0;
Sprite.TARGET = 1;
Sprite.BOX2 = 2;
Sprite.DEFAULT_FRAME_RATE = 150;

Sprite.SPRITE_WIDTH = [27, 45, 45];
Sprite.SPRITE_HEIGHT = [45, 44, 44];
Sprite.SPRITE_SCAN = [1, 1, 1];
Sprite.SPRITE_OFFSET_X = [-14, -21, -21];
Sprite.SPRITE_OFFSET_Y = [-41, -32, -32];
Sprite.FRAMES_MOVING = [1, 0, 2, 0];
